File: 3. Behavioral Design Patterns/Visitor/Go/main.go
package main

import (
	"fmt"
	"strings"
)

// Visitor Pattern Example in Go
// Document structure with different element types processed by visitors

type DocumentVisitor interface {
	VisitParagraph(*Paragraph)
	VisitImage(*Image)
	VisitTable(*Table)
}

type DocumentElement interface {
	Accept(DocumentVisitor)
}

type Paragraph struct {
	Text string
}

func (p *Paragraph) Accept(visitor DocumentVisitor) {
	visitor.VisitParagraph(p)
}

type Image struct {
	Filename string
}

func (i *Image) Accept(visitor DocumentVisitor) {
	visitor.VisitImage(i)
}

type Table struct {
	Rows    int
	Columns int
}

func (t *Table) Accept(visitor DocumentVisitor) {
	visitor.VisitTable(t)
}

type WordCountVisitor struct {
	WordCount int
}

func (w *WordCountVisitor) VisitParagraph(p *Paragraph) {
	words := strings.Fields(p.Text)
	w.WordCount += len(words)
}

func (w *WordCountVisitor) VisitImage(i *Image) {
	// images don't contribute to word count
}

func (w *WordCountVisitor) VisitTable(t *Table) {
	w.WordCount += t.Rows * t.Columns // assume 1 word per cell
}

type PrintVisitor struct{}

func (pv *PrintVisitor) VisitParagraph(p *Paragraph) {
	fmt.Printf("Paragraph: %s\n", p.Text)
}

func (pv *PrintVisitor) VisitImage(i *Image) {
	fmt.Printf("Image: %s\n", i.Filename)
}

func (pv *PrintVisitor) VisitTable(t *Table) {
	fmt.Printf("Table: %dx%d\n", t.Rows, t.Columns)
}

func main() {
	elements := []DocumentElement{
		&Paragraph{"Hello world from paragraph one."},
		&Image{"photo.jpg"},
		&Table{3, 4},
		&Paragraph{"This is another paragraph."},
	}

	wordCounter := &WordCountVisitor{}
	printer := &PrintVisitor{}

	for _, element := range elements {
		element.Accept(wordCounter)
		element.Accept(printer)
	}

	fmt.Printf("Total word count: %d\n", wordCounter.WordCount)
}

File: 3. Behavioral Design Patterns/Visitor/Rust/main.rs
// Visitor Pattern Example in Rust
// Document structure with different element types processed by visitors

trait DocumentVisitor {
    fn visit_paragraph(&mut self, paragraph: &Paragraph);
    fn visit_image(&mut self, image: &Image);
    fn visit_table(&mut self, table: &Table);
}

trait DocumentElement {
    fn accept(&self, visitor: &mut dyn DocumentVisitor);
}

struct Paragraph {
    text: String,
}

impl Paragraph {
    fn new(text: &str) -> Self {
        Self { text: text.to_string() }
    }
}

impl DocumentElement for Paragraph {
    fn accept(&self, visitor: &mut dyn DocumentVisitor) {
        visitor.visit_paragraph(self);
    }
}

struct Image {
    filename: String,
}

impl Image {
    fn new(filename: &str) -> Self {
        Self { filename: filename.to_string() }
    }
}

impl DocumentElement for Image {
    fn accept(&self, visitor: &mut dyn DocumentVisitor) {
        visitor.visit_image(self);
    }
}

struct Table {
    rows: usize,
    columns: usize,
}

impl Table {
    fn new(rows: usize, columns: usize) -> Self {
        Self { rows, columns }
    }
}

impl DocumentElement for Table {
    fn accept(&self, visitor: &mut dyn DocumentVisitor) {
        visitor.visit_table(self);
    }
}

struct WordCountVisitor {
    word_count: usize,
}

impl WordCountVisitor {
    fn new() -> Self {
        Self { word_count: 0 }
    }

    fn get_word_count(&self) -> usize {
        self.word_count
    }
}

impl DocumentVisitor for WordCountVisitor {
    fn visit_paragraph(&mut self, paragraph: &Paragraph) {
        self.word_count += paragraph.text.split_whitespace().count();
    }

    fn visit_image(&mut self, _image: &Image) {
        // images don't contribute to word count
    }

    fn visit_table(&mut self, table: &Table) {
        self.word_count += table.rows * table.columns; // assume 1 word per cell
    }
}

struct PrintVisitor;

impl DocumentVisitor for PrintVisitor {
    fn visit_paragraph(&mut self, paragraph: &Paragraph) {
        println!("Paragraph: {}", paragraph.text);
    }

    fn visit_image(&mut self, image: &Image) {
        println!("Image: {}", image.filename);
    }

    fn visit_table(&mut self, table: &Table) {
        println!("Table: {}x{}", table.rows, table.columns);
    }
}

fn main() {
    let elements: Vec<Box<dyn DocumentElement>> = vec![
        Box::new(Paragraph::new("Hello world from paragraph one.")),
        Box::new(Image::new("photo.jpg")),
        Box::new(Table::new(3, 4)),
        Box::new(Paragraph::new("This is another paragraph.")),
    ];

    let mut word_counter = WordCountVisitor::new();
    let mut printer = PrintVisitor;

    for element in &elements {
        element.accept(&mut word_counter);
        element.accept(&mut printer);
    }

    println!("Total word count: {}", word_counter.get_word_count());
}
